x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/DatafeedConfigReader.java

```diff
@@ -41,6 +41,28 @@ public DatafeedConfigReader(DatafeedConfigProvider datafeedConfigProvider) {
         this.datafeedConfigProvider = datafeedConfigProvider;
     }
 
+    /**
+     * Read the datafeed config from {@code state} and if not found
+     * look for the index document
+     *
+     * @param datafeedId Id of datafeed to get
+     * @param state      Cluster state
+     * @param listener   DatafeedConfig listener
+     */
+    public void datafeedConfig(String datafeedId, ClusterState state, ActionListener<DatafeedConfig> listener) {
+        MlMetadata mlMetadata = MlMetadata.getMlMetadata(state);
+        DatafeedConfig config = mlMetadata.getDatafeed(datafeedId);
+
+        if (config != null) {
+            listener.onResponse(config);
+        } else {
```

Comment:
Why isn't this the other way around? I would think that since we are moving towards putting things in the index, that we would want that to be our positive path, and we would always check it first and return it if we find it.

This concern is also for https://github.com/elastic/elasticsearch/pull/35598/files#diff-db8b21d42edc9e4fada4630d49530113R141

Reply:
True I haven't been entirely consistent with this. The check for the config in cluster state doesn't require an async call so it makes sense to do the fast check first, then the order of the branching is based on what thoughts I could manage to keep in my brain. Note these code paths are only in 6.6 and 6.7 and will be maintained until 6.7.final_ever_bug_release